compiler/AST/foralls.cpp

```diff
@@ -230,3 +234,763 @@ bool astUnderFI(const Expr* ast, ForallIntents* fi) {
+    if (CallExpr* currCE = toCallExpr(curr))
+      if (currCE->isNamed("chpl_here_free"))
+        { freeHere = curr; break; }
+  }
+  // freeHere==NULL for when riSpec is a user-defined/managed object
+  if (freeHere)
+    freeHere->insertAfter(freeIt->remove());
+}  
+
+void lowerForallStmts() {
+  forv_Vec(ForallStmt, fs, gForallStmts) {
+    if (!fs->inTree() || !fs->getFunction()->isResolved())
+      continue;
+
+    // formerly nonLeaderParCheckInt()
+    FnSymbol* parent = fs->getFunction();
+    // If isTaskFun(parent), error is still reported in nonLeaderParCheckInt.
+    if (parent->isIterator() && !parent->hasFlag(FLAG_INLINE_ITERATOR))
+      USR_FATAL_CONT(fs, "invalid use of parallel construct in serial iterator");
+
+    // Forall intents aka fs->withClause() should have already been handled.
+
+    CallExpr* parIterCall = toCallExpr(fs->firstIterExpr());
+    INT_ASSERT(parIterCall && !parIterCall->next); // expected
+    SET_LINENO(parIterCall);
+    BlockStmt* PARBlock = new BlockStmt();
+    fs->replace(PARBlock); // so we can resolve PARBlock below
+    // Maybe move this ->replace() down to resolveBlockStmt(PARBlock) ?
+    // If so, add ->remove() to fs->iterExprs() and perhaps others.
+
+    // From the original buildStandaloneForallLoopStmt(), with "sa" -> "par".
+    VarSymbol* iterRec = newTemp("chpl__iterPAR"); // serial iter, PAR case
+    VarSymbol* parIter = newTemp("chpl__parIter");
+    VarSymbol* parIdx  = fs->parIdxVar();
+
+    iterRec->addFlag(FLAG_CHPL__ITER);
+    iterRec->addFlag(FLAG_MAYBE_REF);
+    iterRec->addFlag(FLAG_EXPR_TEMP);
+
+    parIter->addFlag(FLAG_EXPR_TEMP);
+    // Too late to do it here - it's needed in setConstFlagsAndCheckUponMove().
+    //parIdx->addFlag(FLAG_INDEX_OF_INTEREST);
+    parIdx->addFlag(FLAG_INDEX_VAR);
+
+    PARBlock->insertAtTail(new DefExpr(iterRec));
+    PARBlock->insertAtTail(new DefExpr(parIter));
+    DefExpr* parIdxDef = toDefExpr(fs->masterBlock()->body.head);
+    INT_ASSERT(parIdxDef && parIdxDef->sym == parIdx);
+    PARBlock->insertAtTail(parIdxDef->remove());
+
+    PARBlock->insertAtTail(new CallExpr(PRIM_MOVE, iterRec, parIterCall));
+    PARBlock->insertAtTail("'move'(%S, _getIterator(%S))", parIter, iterRec);
+    PARBlock->insertAtTail("{TYPE 'move'(%S, iteratorIndex(%S)) }", parIdx, parIter);
+
+    currentAstLoc = fs->loopBody()->astloc; // can't do SET_LINENO
+    ForLoop* PARBody = new ForLoop(parIdx, parIter, NULL, false);
+
+    // not needed:
+    //destructureIndices(PARBody, indices, new SymExpr(parIdxCopy), false);
+
+    PARBlock->insertAtTail(PARBody);
+    PARBlock->insertAtTail("_freeIterator(%S)", parIter);
+
+    resolveBlockStmt(PARBlock);
+    move_freeIterator_call(PARBlock);
+    PARBlock->flattenAndRemove();
+
+    fs->loopBody()->flattenAndRemove(); // into fs->masterBlock()
+
+    PARBody->insertAtTail(fs->masterBlock()); // m.Block is already resolved
+    fs->masterBlock()->flattenAndRemove();    // into PARBody
+  }
+}
```

Comment:
I'd like to see lowerForallStmts and its supporting functions moved to its own file in resolution/.